Add Eslint plugin and no-empty-strings rule

alexandernanberg opened the pull request:
Adds Eslint plugin with `no-empty-strings` rule which checks for empty `desc` and `<fbt></fbt>`



<img width="686" alt="image" src="https://github.com/user-attachments/assets/13212c4b-b024-4748-8715-37a237a53f4c" />

Feel free to bikeshed on the naming! 

Will do follow up PRs for adding more rules

alexandernanberg commented on .eslintrc.cjs:
```diff
@@ -1,7 +1,9 @@
+/* eslint-disable sort-keys-fix/sort-keys-fix */
```
this rule is a bit annoying for certain config object, e.g. here when `root` is usually at the top

cpojer replied:
There is no need for root to be at the top though, is it?

alexandernanberg replied:
No, but it's the common convention from what I've seen. But we didn't need this change so I removed it altogether instead

alexandernanberg commented on packages/eslint-plugin-fbtee/src/__tests__/no-empty-strings-test.ts:
```diff
@@ -0,0 +1,197 @@
+/* eslint-disable import/no-unresolved */
```
Not sure why the import plugin doesn't like `@typescript-eslint/rule-tester`

cpojer replied:
I'll take a look at that tomorrow (Friday).

alexandernanberg commented:
Hmm maybe it's better to split checking the `desc` from the actual content. Seems like what Meta did internally https://github.com/facebook/fbt/issues/329

cpojer commented:
Just fyi, this is what I am using for Athena Crisis at the moment:

https://github.com/nkzw-tech/athena-crisis/blob/main/eslint-plugin/require-fbt-description.js

Note, I also have this plugin to ensure `import { fbt } from 'fbt'` instead of `import fbt from 'fbt'` for functional usage of fbt, but it may no longer be needed with fbtee , due to better typing, see https://github.com/nkzw-tech/athena-crisis/blob/main/eslint-plugin/no-fbt-import.js. I will get back on this one after extensive testing).

cpojer commented:
Can you add a test to make sure it doesn't crash for cases like this one?

https://github.com/nkzw-tech/fbtee/blob/main/packages/fbtee/src/intlList.tsx#L77-L82

cpojer commented on packages/eslint-plugin-fbtee/src/__tests__/no-empty-strings-test.ts:
```diff
@@ -0,0 +1,197 @@
+/* eslint-disable import/no-unresolved */
+/* eslint-disable sort-keys-fix/sort-keys-fix */
```
Most object keys are sorted already. Why disable it in this file?

fyi, I have VS Code setup to auto save as I type, and then`cmd+s` applies fixes so I just write my objects however I want and they are sorted automatically when I hit `cmd+s`. See https://cpojer.net/posts/the-perfect-development-environment#vs-code

alexandernanberg replied:
For this particular case I thought it made sense to have valid tests before invalid ones, but we can change if you prefer 

Nice, I'll check it out! I have prettier formatting on save but not eslint fixes atm, it caused some unexpected changes sometimes with hook dependencies

cpojer commented:
Do you mind only using `tsx` files? I'm trying to stick to one file extension only as much as possible since it's always frustrating when you need JSX in a `ts` file and they have to rename. It then becomes more annoying to go back through git history etc.

I'll take a look at the disabled lint rules tomorrow as I'm out of town atm.

alexandernanberg commented:
Changed all files to `.tsx`

Also removed the `desc` checks from the `no-empty-strings` rule, think it makes more sense to have it in a separate rule e.g. `unhelpful-desc`. Then we can also check that it doesn't equals the text content and is too short like described in https://github.com/facebook/fbt/issues/329

cpojer commented:
Awesome, thank you!
